topdown/cache.go

```diff
@@ -171,3 +171,36 @@ func (e *baseCacheElem) set(value ast.Value) {
 	e.value = value
 	e.children = map[ast.Value]*baseCacheElem{}
 }
+
+type refStack struct {
+	sl []refStackElem
+}
+
+type refStackElem struct {
+	refs []ast.Ref
+}
+
+func newRefStack() *refStack {
+	return &refStack{}
+}
+
+func (s *refStack) Push(refs []ast.Ref) {
+	s.sl = append(s.sl, refStackElem{refs: refs})
+}
+
+func (s *refStack) Pop() {
+	s.sl = s.sl[:len(s.sl)-1]
+}
+
+func (s *refStack) Prefixed(ref ast.Ref) bool {
```

Comment:
Are there any performance implications of replacing the trie implementation ?

Reply:
The check will be O(number of with targets in-scope) so I'm not too worried about it. If it becomes a problem we can revisit ti.